composer/utils/batch_helpers.py

```diff
@@ -0,0 +1,145 @@
+from typing import Any, Sequence
+
+__all__ = ['batch_get', 'batch_set']
+
+
+def batch_get(batch: Any, key: Any) -> Any:
+    """Indexes into the batch given the key.
+    Args:
+        batch (Any): An object that contains the input and label of the items in the batch.
+            Can be any abritrary type that user creates, but we assume some sort of
+            sequence (list, tuple, tensor, array), mapping (dictionary),
+            or attribute store (object with data members, namedtuple).
+        key (BatchKeys): A key to index into the batch. BatchKeys is of type Union[BatchKey Sequence[BatchKey]]
+            and BatchKey is of type Union[int, str, slice].
+
+    Returns:
+        The part of the batch specified by the key. This could be any type depending on
+            what the batch is composed of.
+    """
+    if isinstance(key, Sequence) and not isinstance(key, str):
+        return _batch_get_multiple(batch, key)
+    else:
+        return _batch_get(batch, key)
+
+
+def _batch_get(batch: Any, key: Any) -> Any:
+    if isinstance(batch, tuple):
+        return _batch_get_tuple(batch, key)
+    else:
+        exceptions = []
+        try:
+            value = batch[key]
+
+        # The only acceptable TypeError is for an object that doesn't have a __getitem__,
+        # which is TypeError("... object is not subscriptable").
+        except TypeError as e:
+            if 'object is not subscriptable' in str(e):
+                exceptions.append(e)
+                pass
+            else:
+                raise e
+        else:
+            return value
+
+        try:
+            value = getattr(batch, key)
+
+        # If both getattr and __getitem__ result in exceptions then raise both of them.
+        except Exception as e:  # AttributeError, TypeError.
+            exceptions.append(e)
+            raise Exception(exceptions)
+        else:
+            return value
+
+
+def _batch_get_multiple(batch: Any, key: Any):
+    return [_batch_get(batch, k) for k in key]
+
+
+def batch_set(batch: Any, key: Any, value: Any) -> Any:
+    """Indexes into the batch given the key and sets the element at that index to value.
```

Comment:
should clarify that this is not garaunteed to be an in-place operation (eg. for tuples).

Reply:
ok fixed here: 
2136dcb